common/src/main/java/org/conscrypt/Hkdf.java

```diff
@@ -0,0 +1,231 @@
+     * HMAC algorithm to obtain from the specified Provider.
+     * <p>
+     * The Hkdf instance will try and obtain Mac instances from the specified Provider which
+     * does not need to be installed as a JVM-wide Provider.
+     *
+     * @param hmacName the name of the HMAC algorithm to use
+     * @param provider the Provider to use
+     * @throws NoSuchAlgorithmException if hmacName is not a valid HMAC name
+     */
+    public Hkdf(String hmacName, Provider provider) throws NoSuchAlgorithmException {
+        Objects.requireNonNull(hmacName);
+        Objects.requireNonNull(provider);
+        this.hmacName = hmacName;
+        this.provider = provider;
+
+        // Stash the MAC length with the bonus that we'll fail fast here if no such algorithm.
+        macLength = Mac.getInstance(hmacName, provider).getMacLength();
+    }
+
+    private static Provider findProvider(String requestedName) throws NoSuchProviderException {
+        if (requestedName != null) {
+            Provider provider = Security.getProvider(requestedName);
+            if (provider != null) {
+                return provider;
+            }
+            throw new NoSuchProviderException();
+        }
+        // If no Provider name was supplied, look for possible Conscrypt Providers
+        // in order of preference.
+        List<String> candidates = Arrays.asList(
+            Platform.getDefaultProviderName(), // Our own platform's Provider name
+            "Conscrypt",                       // Android unbundled or OpenJDK
+            "GmsCore_OpenSSL",                 // Google Play Services
+            "AndroidOpenSSL");                 // Android platform
+        for (String candidate : candidates) {
+            Provider provider = Security.getProvider(candidate);
+            if (provider != null) {
+                return provider;
+            }
+        }
+        // No user requested Provider and no Conscrypt Provider installed so instantiate one
+        // for our private use but don't install it.
+        return Conscrypt.newProvider();
+    }
+
+    // Visible for testing.
+    Provider getProvider() {
+        return provider;
+    }
+
+    // Visible for testing.
+    int getMacLength() {
+        return macLength;
+    }
+
+    /**
+     * Performs an HKDF extract operation as specified in RFC 5869.
+     *
+     * @param salt the salt to use
+     * @param ikm initial keying material
+     * @return a pseudorandom key suitable for use in expand operations
+     * @throws InvalidKeyException if the salt is not suitable for use as an HMAC key
+     */
+
+    public byte[] extract(byte[] salt, byte[] ikm) throws InvalidKeyException {
+        Objects.requireNonNull(salt);
+        Objects.requireNonNull(ikm);
+        Preconditions.checkArgument(ikm.length > 0, "Empty keying material");
+        if (salt.length == 0) {
+            salt = new byte[getMacLength()];
```

Comment:
Guessing this comes from "if not provided, it is set to a string of HashLen zeros", and you're saying that length zero is "not provided"?

FWIW, BoringSSL's implementation just says it's always provided and if you pass in an empty string, cool you wanted an empty string. Except this is all moot because HMAC zero-pads all inputs up to the hash's block size (not the same as hash size). So as long as hash size <= block size, which is true for all our hashes, empty string and `new byte[getMacLength()]` do the same thing.

Reply:
Partly that, partly the Mac expects to be initialised with a key, which in practical terms means a RAW SecretKeySpec... And SecretKeySpec throws on zero length keys (other short lengths are fine).